Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp

```diff
@@ -0,0 +1,173 @@
+        result.data = WTFMove(data);
+    }, [&](RefPtr<JSC::ArrayBuffer>&& data) {
+        result.data = WTFMove(data);
+    }, [&](RefPtr<Blob>&& data) {
+        result.data = WTFMove(data);
+    }, [&](String&& data) {
+        result.data = WTFMove(data);
+    });
+
+    return result;
+}
+
+static void fetchDataBytesForWrite(const FileSystemWritableFileStream::DataVariant& data, CompletionHandler<void(bool, std::span<const uint8_t>)>&& completionHandler)
+{
+    WTF::switchOn(data, [&](const RefPtr<JSC::ArrayBufferView>& bufferView) {
+        if (!bufferView || bufferView->isDetached())
+            return completionHandler(false, { });
+
+        RefPtr buffer = bufferView->possiblySharedBuffer();
+        if (!buffer)
+            return completionHandler(false, { });
+
+        completionHandler(true, buffer->span());
+    }, [&](const RefPtr<JSC::ArrayBuffer>& buffer) {
+        if (!buffer || buffer->isDetached())
+            return completionHandler(false, { });
+
+        completionHandler(true, buffer->span());
+    }, [&](const RefPtr<Blob>& blob) {
+        if (!blob)
+            return completionHandler(false, { });
+
+        blob->getArrayBuffer([completionHandler = WTFMove(completionHandler)](auto&& result) mutable {
+            if (result.hasException())
+                return completionHandler(false, { });
+
+            Ref buffer = result.releaseReturnValue();
+            if (buffer->isDetached())
+                return completionHandler(false, { });
+
+            completionHandler(true, buffer->span());
+        });
+    }, [&](const String& string) {
+        completionHandler(true, string.utf8().span());
+    });
+}
+
+ExceptionOr<Ref<FileSystemWritableFileStreamSink>> FileSystemWritableFileStreamSink::create(FileSystemFileHandle& source)
+{
+    return adoptRef(*new FileSystemWritableFileStreamSink(source));
+}
+
+FileSystemWritableFileStreamSink::FileSystemWritableFileStreamSink(FileSystemFileHandle& source)
+    : m_source(source)
+{
+}
+
+FileSystemWritableFileStreamSink::~FileSystemWritableFileStreamSink()
+{
+    if (!m_isClosed)
+        protectedSource()->closeWritable(false);
+}
+
+// https://fs.spec.whatwg.org/#write-a-chunk
+void FileSystemWritableFileStreamSink::write(ScriptExecutionContext& context, JSC::JSValue value, DOMPromiseDeferred<void>&& promise)
+{
+    if (m_isClosed)
+        return promise.reject(Exception { ExceptionCode::InvalidStateError, "Sink is closed"_s });
+
+    if (value.isUndefined())
+        return promise.reject(Exception { ExceptionCode::TypeError, "Data is undefined"_s });
+
+    auto scope = DECLARE_THROW_SCOPE(context.vm());
+    auto chunkResult = convert<IDLUnion<IDLArrayBufferView, IDLArrayBuffer, IDLInterface<Blob>, IDLUSVString, IDLDictionary<FileSystemWritableFileStream::WriteParams>>>(*context.globalObject(), value);
+    if (UNLIKELY(chunkResult.hasException(scope)))
+        return promise.reject(Exception { ExceptionCode::ExistingExceptionError });
```

Comment:
Do we have a test?

Reply:
Will add a test (wpt may have the tests but currently they all fail since this is just a partial implementation).